### DISCUSSION TITLE
Question about finite state machine synchronisde by a clock

### DISCUSSION THREAD

**Discussion by kmerbcn** [modelica/ModelicaSpecification#3535] [General]
Hello

I have the following modelica clock. I define two states inside a model.
I would like to know whats wrong in the following model

 model ClassicalDAB
  parameter Real Ts;
    inner Boolean[4] H1;
    
    block State1
      outer output Boolean[4] H1;
    equation
      H[0] = True;
      H[1] = True;
      H[2] = False;  
      H[3] = False;
    end State1;
    
    block State2
      outer output Boolean[4] H1;
     equation
      H[0] = False;
      H[1] = True;
      H[2] = False;
      H[3] = False;
    end State2;
    
  equation
    initialState(State1);
    transition(State1, State2, sample(0, Ts), immediate=false);
    transition(State2, State1, sample(0, Ts), immediate=false);
end ClassicalDAB;

The error is as follows :

[51] 15:51:07 Translation Error
[CargadorInalambrico: 274:5-274:63]: [Expected State2 to be a component, but found class instead.](omeditmessagesbrowser:///CargadorInalambrico?lineNumber=274)

[1] 15:51:14 Translation Error
[CargadorInalambrico: 272:5-272:25]: [Expected State1 to be a component, but found class instead.](omeditmessagesbrowser:///CargadorInalambrico?lineNumber=272)

What is wrong with the definition? I followed the example given by Mr.  Elmqvist . I would like also to know if the condition of transition given by sampling the clock is also correct. My idea is to do the transition at the edge of every clock.

Regards,

Javier

**Comment by kmerbcn** [comment]
Hello to all,

I am still stuck with the definition of the finite state machines in modelica. I copied and paste the state machine example given by the help and I have the same error as above. When the model is flattened, it appears 2 equations and 1 variable.
Code is below

model PruebaEstados
  inner Integer i(start=0);
  
  block State1
    outer output Integer i;
   equation
    i = previous(i) + 2;
  end State1;
  
  block State2
    outer output Integer i;
   equation
    i = previous(i) - 1;
  end State2;
  
  State1 state1;
  State2 state2;
  
equation
  initialState(state1);
  transition(state1, state2, i > 10, immediate =false ,reset=true , synchronize =false ,priority =1);
  transition(state2, state1, i < 1, immediate =false ,reset=true , synchronize =false ,priority =1);

end PruebaEstados;

What am I doing wrong?

**Comment by HansOlsson** [comment]
I don't see any error with the second example, it could be an issue with the Modelica tool you are using.

The first model should be written as:
```
model ClassicalDAB
  parameter Real Ts=0.1;
  Clock c=Clock(Ts);
  inner Boolean[4] H;
  block State1
    outer output Boolean[4] H;
  equation 
    H[1] =  true;
    H[2] = true;
    H[3] = false;
    H[4] = false;
  end State1;
  State1 state1;
  block State2
    outer output Boolean[4] H;
  equation 
    H[1] = false;
    H[2] = true;
    H[3] = false;
    H[4] = false;
  end State2;
  State2 state2;
equation 
  initialState(state1);
  transition(
    state1,
    state2,
    sample(true, c),
    immediate=false);
  transition(
    state2,
    state1,
    true,
    immediate=false);
end ClassicalDAB;
```
Changes:
- Have both State-class and state-component
- Indices start at 1 not 0, and true/false not True/False
- You should not use `sample(0,Ts)` for clocked system, instead use `sample(..., c);` for a clock `c`.
